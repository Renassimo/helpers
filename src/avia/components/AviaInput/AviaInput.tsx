import { KeyboardEvent, ReactNode } from 'react';

import FormControl from '@mui/material/FormControl';
import Input from '@mui/material/Input';
import InputAdornment from '@mui/material/InputAdornment';
import InputLabel from '@mui/material/InputLabel';

import IconButton from '@mui/material/IconButton';
import CloseIcon from '@mui/icons-material/Close';

const AviaInput = ({
  title,
  placeholder,
  label,
  value,
  setValue,
  disabled,
  onKeyDown,
  adornments,
}: {
  title: string;
  placeholder?: string;
  label?: string;
  value: string;
  setValue: (value: string) => void;
  disabled: boolean;
  onKeyDown?: (event: KeyboardEvent) => void;
  adornments: ReactNode;
}) => {
  return (
    <FormControl fullWidth variant="standard">
      {label && <InputLabel>{label}</InputLabel>}
      <Input
        value={value}
        onChange={(event) => setValue(event.target.value)}
        size="small"
        disabled={disabled}
        placeholder={placeholder ?? title}
        aria-label={title}
        fullWidth
        onKeyDown={onKeyDown}
        endAdornment={
          <InputAdornment position="end">
            <>
              {adornments}
              {value && (
                <IconButton
                  size="small"
                  aria-label="Clear"
                  onClick={() => setValue('')}
                  disabled={disabled}
                  edge="end"
                >
                  {<CloseIcon />}
                </IconButton>
              )}
            </>
          </InputAdornment>
        }
      />
    </FormControl>
  );
};

export default AviaInput;
