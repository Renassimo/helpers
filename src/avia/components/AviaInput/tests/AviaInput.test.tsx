import userEvent from '@testing-library/user-event';
import renderWithTheme from '@/common/tests/helpers/renderWithTheme';

import { KeyboardEvent, ReactNode } from 'react';

import AviaInput from '../AviaInput';

describe('AviaInput', () => {
  const renderAviaInput = (props: {
    title: string;
    placeholder?: string;
    label?: string;
    value: string;
    setValue: (value: string) => void;
    disabled: boolean;
    onKeyDown: (event: KeyboardEvent) => void;
    adornments: ReactNode;
  }) => renderWithTheme(<AviaInput {...props} />);

  const setValue = jest.fn();
  const onKeyDown = jest.fn();

  const additionalProps = {
    placeholder: 'Placeholder',
    label: 'Label',
  };

  const defaultProps = {
    title: 'Title',
    value: 'Value',
    setValue,
    disabled: false,
    onKeyDown,
    adornments: <div>Adornments</div>,
  };

  const fullProps = {
    ...defaultProps,
    ...additionalProps,
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders successfully', () => {
    // Arange
    // Act
    const { baseElement } = renderAviaInput(defaultProps);
    // Assert
    expect(baseElement).toMatchSnapshot();
  });

  describe('when empty value passed', () => {
    test('renders successfully', () => {
      // Arange
      // Act
      const { baseElement } = renderAviaInput({ ...fullProps, value: '' });
      // Assert
      expect(baseElement).toMatchSnapshot();
    });
  });

  describe('when full props passed', () => {
    test('renders successfully', () => {
      // Arange
      // Act
      const { baseElement } = renderAviaInput(fullProps);
      // Assert
      expect(baseElement).toMatchSnapshot();
    });
  });

  describe('when updates value', () => {
    test('calls setValue', async () => {
      // Arange
      const { getByLabelText } = renderAviaInput(defaultProps);
      // Act
      await userEvent.type(getByLabelText('Title'), '-');
      // Assert
      expect(setValue).toHaveBeenCalledWith('Value-');
      expect(onKeyDown).toHaveBeenCalled();
    });
  });

  describe('when clicks clear', () => {
    test('calls setValue', async () => {
      // Arange
      const { getByLabelText } = renderAviaInput(defaultProps);
      // Act
      await userEvent.click(getByLabelText('Clear'));
      // Assert
      expect(setValue).toHaveBeenCalledWith('');
    });
  });

  describe('when disabled passed', () => {
    const props = {
      ...defaultProps,
      disabled: true,
    };

    test('renders successfully', () => {
      // Arange
      // Act
      const { baseElement } = renderAviaInput(props);
      // Assert
      expect(baseElement).toMatchSnapshot();
    });

    describe('when tries to update value', () => {
      test('does not call setValue', async () => {
        // Arange
        const { getByLabelText } = renderAviaInput(props);
        // Act
        await userEvent.type(getByLabelText('Title'), '-');
        // Assert
        expect(setValue).not.toHaveBeenCalled();
      });
    });
  });
});
