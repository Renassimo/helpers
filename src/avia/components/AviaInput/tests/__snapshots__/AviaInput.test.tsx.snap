// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AviaInput renders successfully 1`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
    >
      <div
        aria-label="Title"
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-953pxc-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-nz481w-MuiInputBase-input-MuiInput-input"
          placeholder="Title"
          type="text"
          value="Value"
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
        >
          <div>
            Adornments
          </div>
          <button
            aria-label="Clear"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeSmall css-1bkxqye-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`AviaInput when disabled passed renders successfully 1`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
    >
      <div
        aria-label="Title"
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary Mui-disabled MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-953pxc-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input Mui-disabled MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-nz481w-MuiInputBase-input-MuiInput-input"
          disabled=""
          placeholder="Title"
          type="text"
          value="Value"
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
        >
          <div>
            Adornments
          </div>
          <button
            aria-label="Clear"
            class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-edgeEnd MuiIconButton-sizeSmall css-1bkxqye-MuiButtonBase-root-MuiIconButton-root"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`AviaInput when empty value passed renders successfully 1`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
      >
        Label
      </label>
      <div
        aria-label="Title"
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-953pxc-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-nz481w-MuiInputBase-input-MuiInput-input"
          placeholder="Placeholder"
          type="text"
          value=""
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
        >
          <div>
            Adornments
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`AviaInput when full props passed renders successfully 1`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
      >
        Label
      </label>
      <div
        aria-label="Title"
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-953pxc-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-nz481w-MuiInputBase-input-MuiInput-input"
          placeholder="Placeholder"
          type="text"
          value="Value"
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
        >
          <div>
            Adornments
          </div>
          <button
            aria-label="Clear"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeSmall css-1bkxqye-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
